### PR TITLE
Update demosaicing.cpp

### LINKED ISSUE #18558: A typo in source code of Edge-aware demosaicing.
https://github.com/opencv/opencv/blob/1b443219ed181124978ed8627fb2da8db42765ab/modules/imgproc/src/demosaicing.cpp#L1587
The code
`D[2] = (S[-sstep-1] + S[-sstep+1] + S[sstep-1] + S[sstep+1]) >> 2;`
should be modifed to
`D[2] = (S[-sstep-1] + S[-sstep+1] + S[sstep-1] + S[sstep+1] + 2) >> 2;`

### PR DESCRIPTION
**Pull request by krush11**
fixes #18558

<cut/>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by krush11** [comment]
fixes #18558

File: modules/imgproc/src/demosaicing.cpp [hunk 1]
```diff
@@ -1584,7 +1584,7 @@ class Bayer2RGB_EdgeAware_T_Invoker :
                 {
                     D[0] = S[0];
                     D[1] = (std::abs(S[-1] - S[1]) > std::abs(S[sstep] - S[-sstep]) ? (S[sstep] + S[-sstep] + 1) : (S[-1] + S[1] + 1)) >> 1;
-                    D[2] = (S[-sstep-1] + S[-sstep+1] + S[sstep-1] + S[sstep+1]) >> 2;
+                    D[2] = (S[-sstep-1] + S[-sstep+1] + S[sstep-1] + S[sstep+1] + 2) >> 2;
```
Comment (alalek) [inline modules/imgproc/src/demosaicing.cpp:1587]: To make all code consistent please fix these lines too:
- 1569
- 1571

Also need to check `SIMDInterpolator` implementation.

**Review by alalek — COMMENTED** [review]
As a "bugfix" this patch should go into [3.4 branch](https://github.com/opencv/opencv/wiki/Branches) first.
We will merge changes from 3.4 into master regularly (weekly/bi-weekly).

File: modules/imgproc/src/demosaicing.cpp [hunk 2]
```diff
@@ -1566,9 +1566,9 @@ class Bayer2RGB_EdgeAware_T_Invoker :
             int x = 1;
             if (start_with_green)
             {
-                D[blue<<1] = (S[-sstep] + S[sstep]) >> 1;
+                D[blue<<1] = (S[-sstep] + S[sstep] + 2) >> 1;
                 D[1] = S[0];
-                D[2-(blue<<1)] = (S[-1] + S[1]) >> 1;
+                D[2-(blue<<1)] = (S[-1] + S[1] + 2) >> 1;
```
Comment (alalek) [inline modules/imgproc/src/demosaicing.cpp:1571]: `+ 1` must be here

Comment (krush11) [reply modules/imgproc/src/demosaicing.cpp:1571]: I did not get you as i am not totally familiar with the code base being a new contributor. Can you elaborate on what you mean by +1 ?

Comment (alalek) [reply modules/imgproc/src/demosaicing.cpp:1571]: You should follow the rounding rules. https://en.wikipedia.org/wiki/Rounding

`>> 1` means `* 0.5` (optimized version)
`>> 2` means `* 0.25` (optimized version)

You are switching from `Rounding towards zero` to `Round half up`.

In the second case you adding 0.5 <==> `(x + 2) >> 2`
In the first case you should add 0.5 too ==> `(x + 1) >> 1`

Comment (krush11) [reply modules/imgproc/src/demosaicing.cpp:1571]: oh ok thanks will make the required changes

**Comment by krush11** [comment]
so do i have to make further changes? or will the PR be merged to 3.4 after clearing all checks?

**Review by alalek — APPROVED** [review]

**Pull request merged**